File: frontend/src/lib/play/narrationSanitize.ts
import type { PendingAction } from '@/types';

const LEAKED_TOOL_BLOCK =
  /<tool_call>\s*(\w+)(.*?)(?:<\/tool_call>|$)/gis;
const ARG_PAIR =
  /<arg_key>\s*([^<]+?)\s*<\/arg_key>\s*<arg_value>\s*(.*?)(?=\s*<arg_key>|<\/tool_call>|$)/gis;
const DANGLING_TOOL_TAIL = /<tool_call>.*$/is;

const INT_FIELDS = new Set(['dice_count', 'modifier', 'dc']);

function parseArgPairs(body: string): Record<string, string> {
  const args: Record<string, string> = {};
  for (const match of body.matchAll(ARG_PAIR)) {
    args[match[1].trim()] = match[2].trim();
  }
  return args;
}

function coerceRollArgs(raw: Record<string, string>): Partial<PendingAction> {
  const out: Record<string, string | number> = {};
  for (const [key, value] of Object.entries(raw)) {
    if (INT_FIELDS.has(key)) {
      const n = Number.parseInt(value, 10);
      out[key] = Number.isNaN(n) ? value : n;
    } else {
      out[key] = value;
    }
  }
  return {
    action_type: 'ask_player_roll',
    tool_call_id: 'recovered-roll',
    purpose: String(out.purpose ?? 'Roll'),
    dice_count: Number(out.dice_count ?? 1),
    dice_type: (out.dice_type as PendingAction['dice_type']) ?? 'd20',
    stat: out.stat != null ? String(out.stat) : undefined,
    modifier: out.modifier != null ? Number(out.modifier) : undefined,
    dc: out.dc != null ? Number(out.dc) : undefined,
    vs_label: out.vs_label != null ? String(out.vs_label) : undefined,
    adv_type: out.adv_type != null ? String(out.adv_type) : undefined,
    adv_reason: out.adv_reason != null ? String(out.adv_reason) : undefined,
    success_text:
      out.success_text != null ? String(out.success_text) : undefined,
    fail_text: out.fail_text != null ? String(out.fail_text) : undefined,
    footer: out.footer != null ? String(out.footer) : undefined,
  };
}

/** Strip leaked tool-call markup; recover ask_player_roll args when present. */
export function extractLeakedAskPlayerRoll(text: string): {
  cleaned: string;
  args: Partial<PendingAction> | null;
} {
  let leakedArgs: Partial<PendingAction> | null = null;

  const cleaned = text
    .replace(LEAKED_TOOL_BLOCK, (_full, tool: string, body: string) => {
      if (tool.toLowerCase() !== 'ask_player_roll' || leakedArgs != null) {
        return '';
      }
      const raw = parseArgPairs(body);
      if (Object.keys(raw).length > 0) {
        leakedArgs = coerceRollArgs(raw);
      }
      return '';
    })
    .replace(DANGLING_TOOL_TAIL, '')
    .trim();

  return { cleaned, args: leakedArgs };
}

export function sanitizeNarrationText(text: string): string {
  return extractLeakedAskPlayerRoll(text).cleaned;
}

File: backend/api/narration_sanitize.py
"""Strip model-leaked tool markup from player-facing narration."""

from __future__ import annotations

import re
from typing import Any

LEAKED_TOOL_BLOCK = re.compile(
    r"<tool_call>\s*(?P<tool>\w+)(?P<body>.*?)(?:</tool_call>|$)",
    re.IGNORECASE | re.DOTALL,
)
ARG_PAIR = re.compile(
    r"<arg_key>\s*(?P<key>[^<]+?)\s*</arg_key>\s*<arg_value>\s*"
    r"(?P<value>.*?)(?=\s*<arg_key>|</tool_call>|$)",
    re.IGNORECASE | re.DOTALL,
)
DANGLING_TOOL_TAIL = re.compile(
    r"<tool_call>.*$",
    re.IGNORECASE | re.DOTALL,
)

_INT_FIELDS = frozenset({"dice_count", "modifier", "dc"})


def _parse_arg_pairs(body: str) -> dict[str, str]:
    args: dict[str, str] = {}
    for match in ARG_PAIR.finditer(body):
        args[match.group("key").strip()] = match.group("value").strip()
    return args


def _coerce_roll_args(raw: dict[str, str]) -> dict[str, Any]:
    out: dict[str, Any] = {}
    for key, value in raw.items():
        if key in _INT_FIELDS:
            try:
                out[key] = int(value)
            except ValueError:
                out[key] = value
        else:
            out[key] = value
    out.setdefault("dice_count", 1)
    out.setdefault("dice_type", "d20")
    return out


def extract_leaked_ask_player_roll(text: str) -> tuple[str, dict[str, Any] | None]:
    """Remove leaked tool-call markup; recover ask_player_roll args when present."""
    leaked_args: dict[str, Any] | None = None

    def repl(match: re.Match[str]) -> str:
        nonlocal leaked_args
        if match.group("tool").lower() != "ask_player_roll" or leaked_args is not None:
            return ""
        raw = _parse_arg_pairs(match.group("body"))
        if raw:
            leaked_args = _coerce_roll_args(raw)
        return ""

    cleaned = LEAKED_TOOL_BLOCK.sub(repl, text)
    cleaned = DANGLING_TOOL_TAIL.sub("", cleaned).strip()
    return cleaned, leaked_args
